docs/tutorial/reduction.rst

```diff
@@ -0,0 +1,720 @@
+
+    T arr[4] = {
+        i + 0 < front_size ? front[i + 0] : zero_elem,
+        i + 1 < front_size ? front[i + 1] : zero_elem,
+        i + 2 < front_size ? front[i + 2] : zero_elem,
+        i + 3 < front_size ? front[i + 3] : zero_elem
+    }
+
+This is performed to make it easier for the compiler to recognize vector loads from global. As the performance at large is dominated by how you move the data, it's only natural to utilize dedicated instructions to move more data with less binary. This is evident by the huge performance improvement when loading two values per thread, See `the compiler explorer <https://godbolt.org/z/b36Eea69q>`_ to learn how loading for AMD (both RDNA and CDNA) compiles to ``global_load_dwordx4``, where ``x4`` denotes the 4-vector variant of the instruction.
+
+.. note::
+
+    Note that ``read_global_safe`` that used to take an ``uint32_t`` as the index type, takes a signed integer now. When indexing an array with unsigned integers, the compiler has to handle integer overflows, as they're defined by the C/C++ standards. It might happen, that some part of the vector load indices overflow, thus resulting in a non-contiguous read. If you change the previously linked code to use an unsigned integer as the thread id, the compiler won't emit a vector load. Signed integer overflow is an undefined behavior, and hence unknown to the optimizer. To convey the absence of overflow to the compiler with unsigned indices, add ``__builtin_assume(gid + 4 > gid)``, or the more portable ``[[assume]](gid + 4 > gid)``, once ``amdclang++`` supports it.
+
+``read_global_safe`` implementation is an Immediately Invoked Lambda Expression (IILE), because ``ItemsPerThread`` is an integer value, while you need a compile-time ``iota``-like sequence of integers as a pack for the fold-expressions to expand on. This can only take place as part of template argument deduction on the IILE.
+
+.. _processing-items:
+
+Processing ``ItemsPerThread``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Once the kernel reads ``ItemsPerThread`` number of inputs to local, it immediately reduces them to a scalar. There is no reason to propagate the input element multiplicity to the warp reduction phase.
+
+.. code-block:: C++
+
+    T res = [&]()
+    {
+        // Read input from front buffer to local
+        hip::static_array<T, ItemsPerThread> arr = read_global_safe(gid);
+
+        // Reduce ItemsPerThread to scalar
+        tmp::static_for<1, tmp::less_than<ItemsPerThread>, tmp::increment<1>>([&]<int I>()
+        {
+            get<0>(arr) = op(get<0>(arr), get<I>(arr));
+        });
+
+        return get<0>(arr);
+    }();
+
+Two-pass reduction
+------------------
+
+Alter kernel launch and input fetching such that no more blocks are launched than what a subsequent kernel launch's single block can conveniently reduce, while performing multiple passes of input reading from global and combining their results before engaging in the end game tree-like reduction.
```

Comment:
Something is off in this sentence. I can't quite understand it. Difficult to comment.